sdk/communication/azure-communication-chat/azure/communication/chat/_shared/user_credential_async.py

```diff
@@ -3,87 +3,119 @@
+                    await self._wait_till_lock_owner_finishes_refreshing()
                 else:
                     should_this_thread_refresh = True
                     self._some_thread_refreshing = True
                     break
 
-
         if should_this_thread_refresh:
             try:
-                newtoken = await self._token_refresher()  # pylint:disable=not-callable
-
+                new_token = await self._token_refresher()
+                if not self._is_token_valid(new_token):
+                    raise ValueError(
+                        "The token returned from the token_refresher is expired.")
                 async with self._lock:
-                    self._token = newtoken
+                    self._token = new_token
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
             except:
                 async with self._lock:
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
-
                 raise
-
+        if self._refresh_proactively:
+            self._schedule_refresh()
         return self._token
 
-    async def _wait_till_inprogress_thread_finish_refreshing(self):
+    def _schedule_refresh(self):
+        if self._timer is not None:
+            self._timer.cancel()
+
+        token_ttl = self._token.expires_on - get_current_utc_as_int()
+
+        if self._is_token_expiring_soon(self._token):
+            # Schedule the next refresh for when it reaches a certain percentage of the remaining lifetime.
+            timespan = token_ttl // 2
+        else:
+            # Schedule the next refresh for when it gets in to the soon-to-expire window.
+            timespan = token_ttl - timedelta(
+                minutes=self._DEFAULT_AUTOREFRESH_INTERVAL_MINUTES).total_seconds()
+
+        self._timer = AsyncTimer(timespan, self._update_token_and_reschedule)
+        self._timer.start()
+
+    async def _wait_till_lock_owner_finishes_refreshing(self):
+
         self._lock.release()
         await self._lock.acquire()
 
-    def _token_expiring(self):
-        return self._token.expires_on - get_current_utc_as_int() <\
-            timedelta(minutes=self._ON_DEMAND_REFRESHING_INTERVAL_MINUTES).total_seconds()
+    def _is_token_expiring_soon(self, token):
+        if self._refresh_proactively:
+            interval = timedelta(
+                minutes=self._DEFAULT_AUTOREFRESH_INTERVAL_MINUTES)
+        else:
+            interval = timedelta(
+                minutes=self._ON_DEMAND_REFRESHING_INTERVAL_MINUTES)
+        return ((token.expires_on - get_current_utc_as_int())
+            < interval.total_seconds())
 
-    def _is_currenttoken_valid(self):
-        return get_current_utc_as_int() < self._token.expires_on
+    @classmethod
+    def _is_token_valid(cls, token):
+        return get_current_utc_as_int() < token.expires_on
 
     async def close(self) -> None:
         pass
```

Comment:
As discussed in the meeting - this should implement the logic currently in `__aexit__` - which should then call this function. (We also need this on the sync credential).